runtime/compiler/env/j9method.cpp

```diff
@@ -6766,30 +6775,52 @@ TR_ResolvedJ9Method::getResolvedDynamicMethod(TR::Compilation * comp, I_32 callS
-      TR::VMAccessCriticalSection getResolvedDynamicMethod(fej9());
+      // "unresolvedInCP" is a bit of a misnomer here, but we can describe
+      // something equivalent by checking the callSites table.
+      //
+      *unresolvedInCP = isUnresolvedCallSiteTableEntry(callSiteIndex);
+      }
 
-      J9Class    *ramClass = constantPoolHdr();
-      J9ROMClass *romClass = romClassPtr();
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   J9InvokeCacheEntry *invokeCache = (J9InvokeCacheEntry *) callSiteTableEntryAddress(callSiteIndex);
 
-      if (unresolvedInCP)
+   if (!isUnresolvedCallSiteTableEntry(callSiteIndex))
+      {
+      TR_OpaqueMethodBlock * targetJ9MethodBlock = NULL;
          {
-         // "unresolvedInCP" is a bit of a misnomer here, but we can describe
-         // something equivalent by checking the callSites table.
-         //
-         *unresolvedInCP = (ramClass->callSites[callSiteIndex] == NULL);
+         TR::VMAccessCriticalSection getResolvedDynamicMethod(fej9());
+         targetJ9MethodBlock = fej9()->targetMethodFromMemberName((uintptr_t) invokeCache->target);
          }
-
-      J9SRP                 *namesAndSigs = (J9SRP*)J9ROMCLASS_CALLSITEDATA(romClass);
-      J9ROMNameAndSignature *nameAndSig   = NNSRP_GET(namesAndSigs[callSiteIndex], J9ROMNameAndSignature*);
-      J9UTF8                *signature    = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
-
-      TR_OpaqueMethodBlock *dummyInvokeExact = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "invokeExact", JSR292_invokeExactSig);
-      result = _fe->createResolvedMethodWithSignature(comp->trMemory(), dummyInvokeExact, NULL, utf8Data(signature), J9UTF8_LENGTH(signature), this);
+      result = fej9()->createResolvedMethod(comp->trMemory(), targetJ9MethodBlock, this);
+      return result;
       }
+#endif
 
+   J9SRP                 *namesAndSigs = (J9SRP*)J9ROMCLASS_CALLSITEDATA(romClass);
+   J9ROMNameAndSignature *nameAndSig   = NNSRP_GET(namesAndSigs[callSiteIndex], J9ROMNameAndSignature*);
+   J9UTF8                *signature    = J9ROMNAMEANDSIGNATURE_SIGNATURE(nameAndSig);
+
+#if defined(J9VM_OPT_OPENJDK_METHODHANDLE)
+   TR_OpaqueMethodBlock *dummyInvoke = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "linkToStatic", "([Ljava/lang/Object;)Ljava/lang/Object;");
+   char * linkToStaticSignature = (char *) comp->trMemory()->allocateMemory((J9UTF8_LENGTH(signature)+40)*sizeof(char), heapAlloc);
+   char * romMethodSignature = (char *) comp->trMemory()->allocateMemory((J9UTF8_LENGTH(signature))*sizeof(char), heapAlloc);
+   strcpy(romMethodSignature, utf8Data(signature));
+   char * sigTokenStart = strtok(romMethodSignature, ")");
+   char * sigTokenEnd = strtok(NULL, ")");
+   sprintf(linkToStaticSignature, "%sLjava/lang/Object;Ljava/lang/Object;)%s", sigTokenStart, sigTokenEnd);
+   result = _fe->createResolvedMethodWithSignature(comp->trMemory(), dummyInvoke, NULL, linkToStaticSignature, strlen(linkToStaticSignature), this);
```

Comment:
You don't need to allocate new memory to store the signature. Like the original code, you can do
```
TR_OpaqueMethodBlock *dummyInvoke = _fe->getMethodFromName("java/lang/invoke/MethodHandle", "linkToStatic", "([Ljava/lang/Object;)Ljava/lang/Object;");
result = _fe->createResolvedMethodWithSignature(comp->trMemory(), dummyInvoke, NULL, utf8Data(signature), J9UTF8_LENGTH(signature), this);
```

Reply:
When the call is unresolved, we create a dummy invoke to `linkToStatic`. `linkToStatic` is signature polymorphic and therefore can have any number of args for the call, and return any type. We also need to provide the `appendix` and `target` objects as the last two args. `utf8Data(signature) will just get us the call site signature without accounting for the 2 objects we are providing as args.` Therefore, we take the ROM method signature, and split it where we need to insert the 2 additional parameters in the signature. `strtok` modifies the strings, which is why we make a writable copy of it to work on. I will add some comments explaining the rationale behind this and also reduce the duplication with `invokeHandle`.